Cython

pschella opened the pull request:
Solution to challenge using Cython
- Passes all basics_test.py
- Passes all converters_test.py

Remaining issues:
- No containers or extensions yet
- Had to change TypeError to Attribute error in unit test (also more Pythonic)
- Uses separate Doodad and ImmutableDoodad types on Python side to represent const
- Build issue requires hard coded path on OSX (stupid, will fix)

Commit:
Add documentation

Commit:
Wrap converters module

Implement Swig typemaps for conversion of a Cython wrapped Doodad
to and from Swig wrapped functions

Commit:
Wrap basics module

Provides Cython wrapped types basics.Doodad and basics.Secret.
Also provides an ImmutableDoodad to represent const Doodad objects.

Commit:
Add distutils installer

Commit:
Wrap containers module

Provides a Cython wrapped type DoodadSet

Commit:
Cleanup tests

- setting an immutable object raises an AttributeError
- fix shadowing of test_pass_reference
- use ImmutableDoodad for const object tests

TallJimbo commented on include/containers.hpp:
```diff
@@ -4,6 +4,7 @@
 #include "basics.hpp"
 #include <vector>
 #include <map>
+#include <iostream>
```
This looks like it's probably unnecessary; a relic of some debugging?

pschella replied:
Yes, very likely. I'll clean it up.

pschella commented:
Very likely indeed. I'll clean it up.

Commit:
Remove unused include

TallJimbo commented on python2-cython/challenge/basics.pyx:
```diff
@@ -0,0 +1,252 @@
+            elif isinstance(name, str):
+                self.thisptr.reset(new _Doodad(name, value))
+            else:
+                raise TypeError("a string or tuple is required")
```
Could you remove the need for the `init` argument somehow?  Seems like an implementation detail the user shouldn't see.  Maybe by implementing `__new__`, and having that take a `thisptr` argument?

pschella replied:
I'd love to. But I haven't found a way how to yet. The problem is that this is one of the cases where Cython is different from Python. Cython had `__new__` before Python came up with it and it meant something different. It performed pre-initialization at the C level (e.g. malloc). Now that Python has `__new__` this has been deprecated and renamed as `__cinit__`. But Cython doesn't have the real Python `__new__`. Or at least it doesn't look like it. When I use `d = Doodad.__new__(Doodad)` to skip `__init__` instead of the ugly `d = Doodad(init=False)` it doesn't know about it's C types and thus the subsequent assignment of the `shared_ptr` tries to cast to a Python object which it can't so the cythonizing fails there.

TallJimbo commented on python2-cython/challenge/basics.pyx:
```diff
@@ -0,0 +1,252 @@
+
+        d.thisptr = move(deref(self.thisptr).clone())
+
+        return d
```
I'd delete the extra whitespace in this function (and others like it).

pschella replied:
Done

TallJimbo commented on python2-cython/challenge/containers.pyx:
```diff
@@ -0,0 +1,112 @@
+        """Return Python list of objects in set
+
+        Note that new Python objects will be created share
+        the same underlying C++ Doodad's stored in the set.
```
"Note that _the_ new Python objects _that_ will be created _will_ share _C++_ _Doodads_ with the set."

pschella replied:
Fixed.

TallJimbo commented on python2-cython/challenge/containers.pyx:
```diff
@@ -0,0 +1,112 @@
+        return results
+
+    cpdef assign(self, seq) except +:
+        """Assign Doodad's to set
```
No apostrophe.

pschella replied:
Fixed.

TallJimbo commented on tests/basics_test.py:
```diff
@@ -90,9 +90,9 @@ def test_const(self):
-        self.assertRaises(TypeError, setattr, d, "value", 9)
+        self.assertIsInstance(d, challenge.basics.ImmutableDoodad)
+        self.assertRaises(AttributeError, setattr, d, "name", "k")
+        self.assertRaises(AttributeError, setattr, d, "value", 9)
```
I don't mind changing the expected exception, but we ought to have some way to make the `isinstance` check work.  Maybe by making both `ImmutableDoodad` and `MutableDoodad` inherit from `Doodad` as an ABC?

That said, I don't know that we want to make that change.  But maybe it'd be good to comment that this situation isn't ideal.

pschella replied:
Ok, I have changed this. Note that this does require even more test modification. But it might be nicer this way. Tell me what you think.

Commit:
Fix whitespace

Commit:
Correct spelling

Commit:
Make Doodad an abstract base class for MutableDoodad and ImmutableDoodad

There are now mutable- and immutable-doodaads which each test
True for isinstance(Doodad).

Commit:
Revert "Make Doodad an abstract base class for MutableDoodad and ImmutableDoodad"
Old solution was deemed better.

This reverts commit 62488fed329dc4fc9fe74cec23d3dfdb72390226.

TallJimbo commented:
I've merged this at 20183ed73f14de596cf2c5150b84c36c23b87beb after making a few more modifications on a derived branch:
 - Modified build system to work on gcc/linux (or at least work for me), by making C++ standard library specification only on OX X.
 - Added tests to the build system.
 - Made Swig converters to accept `Doodad` when C++ wants a `Doodad const` (since conversion should be implicit).
